### PR TITLE
Declaration of cloudevents feature

### PR DESCRIPTION
**Pull request by AnotherDaniel**
As announced in #31, this is making the cloudevents dependency an optional up-rust crate feature.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -17,23 +17,28 @@
...
+//! * `uuid` module, which generates and validates UUIDs as per the uProtocol specification
 //!
 //! For user convenience, all of these modules export their types on up_rust top-level, except for (future) optional features.
 //!
+//! ## Optional features
+//!
+//! Some crate features are made optional, these include:
+//!
+//! * `cloudevents` feature, offering a way to represent uProtocol messages using the [cloudevents data model](https://cloudevents.io/)
+//!
```
Comment (PLeVasseur) [inline src/lib.rs:41]: I'm not sure how "Rust-literate" we're expecting users to be. If we're expecting them to be fairly literate, maybe my comment not needed. Perhaps here we could provide a link to how to enable an optional feature in Rust or provide specifically how to enable that for this crate?

WDYT?

Comment (AnotherDaniel) [reply src/lib.rs:41]: Hm, maybe not in the crates.io description - but I'll add a bit of text to the project README, wdyt?

File: README.md [hunk 2]
```diff
@@ -8,27 +8,28 @@ The module contains the factory methods, serializers, and validators for all dat
...
+Building the library is as simple as running `cargo build` in the project root directory. The build pulls [uprotocol protobuf definitions](https://github.com/eclipse-uprotocol/up-core-api) that constitute the core types used in the library. All of these are compiled automatically by `build.rs`. The resulting Rust code is made available to the project via the `up_core_api::` internal module.
 
-__Note:__ the library current uses protobuf message definitions from the cloudevents project. Specifically, the build process downloads:
+__Note:__ the library uses non-stable features from the uuid crate, notably version 8 UUIDs. These features are defined in `cargo.toml` (where the uuid crate dependency is declared), and require a compiler flag to be included in the build. This is configured in `.cargo/config.toml`.
+
+### Optional features
+
+The up-rust crate contains some optional features, as per the create documentation (refer to src/lib.rs). One of these features is the usage of protobuf message definitions from the cloudevents project. Specifically, with this feature enabled the build process downloads:
```
Comment (PLeVasseur) [inline README.md:19]: `create` => `crate`?

File: README.md [hunk 3]
```diff
@@ -8,27 +8,28 @@ The module contains the factory methods, serializers, and validators for all dat
...
 
-__Note:__ the library uses non-stable features from the uuid crate, notably version 8 UUIDs. These features are defined in `cargo.toml` (where the uuid crate dependency is declared), and require a compiler flag to be included in the build. This is configured in `.cargo/config.toml`.
+```json
+{
+    "rust-analyzer.cargo.features": "all"
+}
+```
 
-### Using the library
+For more details on Rust (cargo) features, please refer to [The Cargo Book](https://doc.rust-lang.org/cargo/reference/features.html).
```
Comment (PLeVasseur) [inline README.md:31]: :heart:

File: src/rpc/rpcmapper.rs [hunk 4]
```diff
@@ -354,6 +359,7 @@ mod tests {
         Ok(message)
     }
 
+    #[cfg(feature = "cloudevents")]
```
Comment (PLeVasseur) [inline src/rpc/rpcmapper.rs:362]: I haven't used feature-gating in Rust very much yet. Did you test this locally to make sure these tests are run w/ the feature on and not when the feature is off?

File: Cargo.toml [hunk 5]
```diff
@@ -33,11 +33,17 @@ repository = "https://github.com/eclipse-uprotocol/up-rust"
 rust-version = "1.66"
 version = "0.1.5"
 
+[features]
+cloudevents = ["dep:cloudevents-sdk"]
```
[the remark this answers is not in the document]
Comment (AnotherDaniel) [reply Cargo.toml]: true

File: Cargo.toml [hunk 6]
```diff
@@ -48,7 +48,7 @@ mediatype = "0.19"
 protobuf = { version = "3.3" }
 rand = { version = "0.8" }
 regex = { version = "1.10" }
-serde = { version = "1.0", features = ["derive"] }
+serde = { version = "1.0", features = ["derive"], optional = true }
```
Comment (sophokles73) [inline Cargo.toml:51]: `serde_json` is also required by the cloudevent-sdk but is not really required by `uurivalidator::tests`. In fact, there is a test which reads in a JSON file that contains valid and invalid URIs ... IMHO this should be replaced by `#[test_case]`s to remove the (unnecessary) dependency on `serde_json`.

File: .github/workflows/lint-and-test.yaml [hunk 7]
```diff
@@ -59,7 +59,7 @@ jobs:
       - name: Run tests and report code coverage
         run: |
           # enable nightly features so that we can also include Doctests
-          RUSTC_BOOTSTRAP=1 cargo tarpaulin -o xml -o lcov -o html --doc --tests
+          RUSTC_BOOTSTRAP=1 cargo tarpaulin -o xml -o lcov -o html --doc --tests --all-features
```
Comment (sophokles73) [inline .github/workflows/lint-and-test.yaml:62]: now we no longer test if it actually compiles/works _without_ the `cloudevents` feature, do we?

Comment (AnotherDaniel) [reply .github/workflows/lint-and-test.yaml:62]: This is test code coverage collection - makes most sense to run with all features.
I'd say we need something dedicated in addition now, to cover the matrix of possible features. Am looking into that atm...

Comment (PLeVasseur) [reply .github/workflows/lint-and-test.yaml:62]: I'll take it you'll ping us when you have a solution for that in mind for another round of code review. :+1:

**Pull request closed without merging**